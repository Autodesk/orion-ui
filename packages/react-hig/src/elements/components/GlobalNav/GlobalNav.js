/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as HIG from 'hig.web';
import * as PropTypes from 'prop-types';

import HIGElement from '../../HIGElement';
import HIGChildValidator from '../../HIGChildValidator';
import createComponent from '../../../adapters/createComponent';

import SideNavComponent, { SideNav } from './SideNav';
import TopNavComponent, { TopNav } from './TopNav';
import SubNavComponent, { SubNav } from './SubNav';
import Slot from './Slot';

class GlobalNav extends HIGElement {
  constructor(initialProps) {
    super(HIG.GlobalNav, initialProps);
  }

  componentDidMount() {
    // Add any children
    if (this.sideNav) {
      this.hig.addSideNav(this.sideNav.hig);
      this.sideNav.componentDidMount();
    }

    if (this.topNav) {
      this.hig.addTopNav(this.topNav.hig);
      this.topNav.componentDidMount();
    }

    if (this.subNav) {
      this.hig.addSubNav(this.subNav.hig);
      this.subNav.componentDidMount();
    }

    if (this.initialProps.sideNavOpen) {
      this.hig.showSideNav();
    } else {
      this.hig.hideSideNav();
    }

    if (this.slot) {
      this.hig.addSlot(this.slot);
    }
  }

  createElement(ElementConstructor, props) {
    switch (ElementConstructor) {
      case TopNav:
        return new TopNav(this.hig.partials.TopNav, props);
      case SubNav:
        return new SubNav(this.hig.partials.SubNav, props);
      case SideNav:
        return new SideNav(this.hig.partials.SideNav, props);
      default:
        throw new Error(`Unknown type ${ElementConstructor.name}`);
    }
  }

  appendChild(instance, beforeChild = {}) {
    if (instance instanceof SideNav) {
      if (this.sideNav) {
        throw new Error('only one SideNav is allowed');
      } else {
        this.sideNav = instance;
        if (this.mounted) {
          this.hig.addSideNav(instance.hig);
          instance.componentDidMount();
        }
      }
    } else if (instance instanceof TopNav) {
      if (this.topNav) {
        throw new Error('only one TopNav is allowed');
      } else {
        this.topNav = instance;
        if (this.mounted) {
          this.hig.addTopNav(instance.hig);
          instance.componentDidMount();
        }
      }
    } else if (instance instanceof SubNav) {
      if (this.SubNav) {
        throw new Error('only one SubNav is allowed');
      } else {
        this.subNav = instance;
        if (this.mounted) {
          this.hig.addSubNav(instance.hig);
          instance.componentDidMount();
        }
      }
    } else {
      throw new Error('unknown type');
    }
  }

  addSlot(element) {
    if (this.mounted) {
      this.hig.addSlot(element);
    } else {
      this.slot = element;
    }
  }

  insertBefore(instance) {
    this.appendChild(instance);
  }

  removeChild(instance) {
    if (instance instanceof SideNav) {
      this.sideNav = null;
    } else if (instance instanceof TopNav) {
      this.topNav = null;
    } else if (instance instanceof SubNav) {
      this.subNav = null;
    }
    instance.unmount();
  }

  commitUpdate(updatePayload, oldProps, newProp) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'sideNavOpen': {
          if (propValue) {
            this.hig.showSideNav();
          } else {
            this.hig.hideSideNav();
          }
          break;
        }
        case 'children': {
          /* no-op */
          break;
        }
        default: {
          console.warn(`${propKey} is unknown`);
        }
      }
    }
  }
}

const GlobalNavComponent = createComponent(GlobalNav);

GlobalNavComponent.propTypes = {
  sideNavOpen: PropTypes.bool,
  children: HIGChildValidator([
    SideNavComponent,
    TopNavComponent,
    SubNavComponent,
    Slot
  ])
};

GlobalNavComponent.__docgenInfo = {
  props: {
    sideNavOpen: {
      description: 'show or hide the SideNav'
    },

    children: {
      description: 'support adding SideNav, SubNav, or TopNav'
    }
  }
};

GlobalNavComponent.SideNav = SideNavComponent;
GlobalNavComponent.TopNav = TopNavComponent;
GlobalNavComponent.SubNav = SubNavComponent;
GlobalNavComponent.Slot = Slot;

export default GlobalNavComponent;
