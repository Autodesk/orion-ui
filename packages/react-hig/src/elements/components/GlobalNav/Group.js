/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';
import HIGNodeList from '../../HIGNodeList';
import HIGChildValidator from '../../HIGChildValidator';
import ModuleComponent, { Module } from './Module';

export class Group extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);
    this.modules = new HIGNodeList({
      type: Module,
      HIGConstructor: this.hig.partials.Module,
      onAdd: (instance, beforeInstance) => {
        this.hig.addModule(instance, beforeInstance);
      }
    });

    this.state = {};
    this._render = this._render.bind(this);
  }

  componentDidMount() {
    this.modules.componentDidMount();
    this._render();
  }

  createElement(ElementConstructor, props) {
    return this.modules.createElement(ElementConstructor, props);
  }

  insertBefore(instance, insertBeforeIndex) {
    this.modules.insertBefore(instance, insertBeforeIndex);
    if (this.mounted) {
      this._render();
    }
  }

  removeChild(instance) {
    this.modules.removeChild(instance);
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    this.processUpdateProps(updatePayload).then(this._render);
  }

  isVisible() {
    return this.state.isVisible;
  }

  expandModules() {
    this.modules.forEach(module => {
      module.commitUpdate({query: undefined})
      module.expandAll();
    });
    this._render();
  }

  collapseModules() {
    this.modules.forEach(module => {
      this.commitUpdate({query: undefined})
      module.collapseAll();
    });
    this._render();
  }

  _render() {
    const matches = this.modules.map(module => {
      module.commitUpdate({
        query: this.props.query,
        activeModule: this.props.activeModule,
        activeSubmodule: this.props.activeSubmodule
      });

      return module.isVisible();
    });

    if (matches.some(m => m)) {
      this.hig.show();
      this.state.isVisible = true;
    } else {
      this.hig.hide();
      this.state.isVisible = false;
    }
  }
}

const GroupComponent = createComponent(Group);

GroupComponent.propTypes = {
  children: HIGChildValidator([ModuleComponent])
};

GroupComponent.__docgenInfo = {
  props: {
    children: {
      description: 'support adding Module'
    }
  }
};

GroupComponent.Module = ModuleComponent;

export default GroupComponent;
