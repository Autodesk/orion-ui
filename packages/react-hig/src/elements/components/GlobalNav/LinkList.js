/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import createComponent from '../../../adapters/createComponent';
import HIGChildValidator from '../../HIGChildValidator';
import HIGNodeList from '../../HIGNodeList';

import LinkComponent, { Link } from './Link';

// Does not extend HIGElement because it's not a real HIG component
export class LinkList {
  constructor(higInstance) {
    this.hig = higInstance;
    this.links = new HIGNodeList({
      type: Link,
      HIGConstructor: this.hig.partials.Link,
      onAdd: (instance, beforeInstance) => {
        this.hig.addLink(instance, beforeInstance);
      }
    });
  }

  mount() {
    this.links.componentDidMount();
  }

  unmount() {
    // no-op
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    // no-op
  }

  createElement(ElementConstructor, props) {
    return this.links.createElement(ElementConstructor, props);
  }

  insertBefore(instance, insertBeforeIndex) {
    this.links.insertBefore(instance, insertBeforeIndex);
  }

  removeChild(instance) {
    this.links.removeChild(instance);
  }
}

const LinkListComponent = createComponent(LinkList);

LinkListComponent.propTypes = {
  children: HIGChildValidator([LinkComponent])
};

LinkListComponent.__docgenInfo = {
  props: {
    children: {
      description: 'support adding Link'
    }
  }
};

LinkListComponent.Link = LinkComponent;

export default LinkListComponent;
