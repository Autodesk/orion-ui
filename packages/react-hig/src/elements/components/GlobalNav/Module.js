/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as PropTypes from 'prop-types';
import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';
import HIGNodeList from '../../HIGNodeList';
import HIGChildValidator from '../../HIGChildValidator';
import SubmoduleComponent, { Submodule } from './Submodule';

export class Module extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);
    this.submodules = new HIGNodeList({
      type: Submodule,
      HIGConstructor: this.hig.partials.Submodule,
      onAdd: (instance, beforeInstance) => {
        this.hig.addSubmodule(instance, beforeInstance);
      }
    });
    this.state = {
      title: initialProps.title
    };
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    const mapping = {
      icon: 'setIcon',
      title: 'setTitle',
      link: 'setLink'
    };

    this.commitUpdateWithMapping(updatePayload, mapping);

    if (Object.keys(updatePayload).includes('title')) {
      this.state.title = updatePayload.title;
    }
  }

  componentDidMount() {
    this.submodules.componentDidMount();
  }

  createElement(ElementConstructor, props) {
    return this.submodules.createElement(ElementConstructor, props);
  }

  insertBefore(instance, insertBeforeIndex) {
    this.submodules.insertBefore(instance, insertBeforeIndex);
  }

  removeChild(instance) {
    this.submodules.removeChild(instance);
  }

  matches(query) {
    return this.state.title.toLowerCase().match(query.toLowerCase());
  }

  filter(query) {
    const childMatches = this.submodules.map(submodule => {
      return submodule.filter(query);
    });

    if (childMatches.some(m => m) || this.matches(query)) {
      this.hig.show();
      return true;
    } else {
      this.hig.hide();
      return false;
    }
  }
}

const ModuleComponent = createComponent(Module);

ModuleComponent.propTypes = {
  icon: PropTypes.string,
  title: PropTypes.string,
  link: PropTypes.string,
  show: PropTypes.func,
  hide: PropTypes.func,
  activate: PropTypes.func,
  deactivate: PropTypes.func,
  showSubmodules: PropTypes.func,
  hideSubmodules: PropTypes.func,
  submodulesClosed: PropTypes.bool,
  addSubmodule: PropTypes.func,
  onClick: PropTypes.func,
  onHover: PropTypes.func,
  children: HIGChildValidator([SubmoduleComponent])
};

ModuleComponent.__docgenInfo = {
  props: {
    icon: {
      description: 'sets the icon of an Module'
    },

    title: {
      description: 'sets the title of an Module'
    },

    link: {
      description: 'sets the link of an Module'
    },

    show: {
      description: 'show (used for filtering)'
    },

    hide: {
      description: 'hide (used for filtering)'
    },

    activate: {
      description: 'activates the module'
    },

    deactivate: {
      description: 'deactivates the module'
    },

    showSubmodules: {
      description: 'show the submodules'
    },

    hideSubmodules: {
      description: 'hide the submodules'
    },

    addSubmodule: {
      description: 'add submodule'
    },

    onClick: {
      description: 'triggered when a link is clicked on'
    },

    onHover: {
      description: 'triggered when a link is hovered over'
    },

    children: {
      description: 'support adding SubModule'
    }
  }
};

ModuleComponent.Submodule = SubmoduleComponent;

export default ModuleComponent;
