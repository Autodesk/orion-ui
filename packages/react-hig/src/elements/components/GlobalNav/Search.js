/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as PropTypes from 'prop-types';
import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';

export class Search extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);

    ['showClearIcon', 'hideClearIcon'].forEach(fn => { this[fn] = this[fn].bind(this) });
  }
  commitUpdate(updatePayload, oldProps, newProps) {
    const mapping = {
      placeholder: 'setPlaceholder',
      query: 'setQuery'
    };

    this.commitUpdateWithMapping(updatePayload, mapping);
  }

  componentDidMount() {
    this.hig.onFocusIn(this.showClearIcon);
    this.hig.onFocusOut(this.hideClearIcon);
  }

  showClearIcon() {
    this.hig.showClearIcon();
  }

  hideClearIcon() {
    this.hig.hideClearIcon();
  }
}

const SearchComponent = createComponent(Search);

SearchComponent.propTypes = {
  query: PropTypes.string,
  placeholder: PropTypes.string,
  showClearIcon: PropTypes.func,
  hideClearIcon: PropTypes.func,
  onClearIconClick: PropTypes.func,
  onInput: PropTypes.func,
  onFocusIn: PropTypes.func,
  onFocusOut: PropTypes.func
};

SearchComponent.__docgenInfo = {
  props: {
    placeholder: {
      description: 'sets the placeholder'
    },

    query: {
      description: 'sets the search query'
    },

    showClearIcon: {
      description: '{func} show the clear input icon'
    },

    hideClearIcon: {
      description: '{func} hide the clear input icon'
    },

    onClearIconClick: {
      description: '{func} triggered when clear icon is clicked'
    },

    onInput: {
      description: 'triggered when search field has something entered'
    },

    onFocusIn: {
      description: 'triggered when search field is in focus'
    },

    onFocusOut: {
      description: 'triggered when search field is not in focus'
    }
  }
};

export default SearchComponent;
