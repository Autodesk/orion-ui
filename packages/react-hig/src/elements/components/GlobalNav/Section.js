/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as PropTypes from 'prop-types';

import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';
import HIGNodeList from '../../HIGNodeList';
import HIGChildValidator from '../../HIGChildValidator';
import GroupComponent, { Group } from './Group';
import CollapseComponent, { Collapse } from './Collapse';

export class Section extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);
    this.groups = new HIGNodeList({
      type: Group,
      HIGConstructor: this.hig.partials.Group,
      onAdd: (instance, beforeInstance) => {
        this.hig.addGroup(instance, beforeInstance);
      }
    });
    this.state = {};
  }

  componentDidMount() {
    this.groups.componentDidMount();
    if (this.collapse) {
      this.hig.addCollapse(this.collapse.hig);
      this.collapse.mount();
    }
  }

  commitUpdate(updatePayload, oldProps, newProp) {
    const mapping = {
      headerLabel: 'setHeaderLabel',
      headerName: 'setHeaderName'
    };

    this.commitUpdateWithMapping(updatePayload, mapping);

    if (Object.keys(updatePayload).includes('query')) {
      this.state.query = updatePayload.query;
      this._render();
    }

    if (Object.keys(updatePayload).includes('expand')) {
      this.state.expand = updatePayload.expand;
      this._render();
    }
  }

  createElement(ElementConstructor, props) {
    switch (ElementConstructor) {
      case Group:
        return this.groups.createElement(ElementConstructor, props);
      case Collapse:
        return new Collapse(this.hig.partials.Collapse, props);
      default:
        throw new Error(`Unknown type ${ElementConstructor.name}`);
    }
  }

  insertBefore(instance, insertBeforeIndex) {
    this.appendChild(instance);
  }

  appendChild(instance) {
    if (instance instanceof Group) {
      this.groups.insertBefore(instance); // calls internal _appendChild if no "before" component
    } else if (instance instanceof Collapse) {
      if (this.collapse) {
        throw new Error('only one Collapse is allowed');
      } else {
        this.collapse = instance;
        if (this.mounted) {
          this.hig.addCollapse(instance.hig);
          instance.mount();
        }
      }
    }
  }

  removeChild(instance) {
    this.groups.removeChild(instance);
  }

  _render() {
    const childVisibility = this.groups.map(group => {
      group.commitUpdate({
        query: this.state.query,
        expanded: this.state.expanded
      });

      return group.isVisible();
    });

    if (childVisibility.some(v => v)) {
      this.hig.show();
    } else {
      this.hig.hide();
    }
  }
}

const SectionComponent = createComponent(Section);

SectionComponent.propTypes = {
  headerLabel: PropTypes.string,
  headerName: PropTypes.string,
  children: HIGChildValidator([GroupComponent, CollapseComponent])
};

SectionComponent.__docgenInfo = {
  props: {
    headerLabel: {
      description: 'sets the label'
    },

    headerName: {
      description: 'sets the name'
    },

    children: {
      description: 'support adding Group and Collapse'
    }
  }
};

SectionComponent.Group = GroupComponent;
SectionComponent.Collapse = CollapseComponent;

export default SectionComponent;
