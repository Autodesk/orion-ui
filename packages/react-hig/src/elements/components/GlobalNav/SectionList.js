/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import createComponent from '../../../adapters/createComponent';
import HIGNodeList from '../../HIGNodeList';
import HIGChildValidator from '../../HIGChildValidator';

import SectionComponent, { Section } from './Section';

// Does not extend HIGElement because it's not a real HIG component
export class SectionList {
  constructor(higInstance) {
    this.hig = higInstance;
    this.sections = new HIGNodeList({
      type: Section,
      HIGConstructor: this.hig.partials.Section,
      onAdd: (instance, beforeInstance) => {
        this.hig.addSection(instance, beforeInstance);
      }
    });
  }

  mount() {
    this.sections.componentDidMount();
  }

  unmount() {
    // no-op
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    // no-op
  }

  createElement(ElementConstructor, props) {
    return this.sections.createElement(ElementConstructor, props);
  }

  insertBefore(instance, insertBeforeIndex) {
    this.sections.insertBefore(instance, insertBeforeIndex);
  }

  removeChild(instance) {
    this.sections.removeChild(instance);
  }
}

const SectionListComponent = createComponent(SectionList);

SectionListComponent.propTypes = {
  children: HIGChildValidator([SectionComponent])
};

SectionListComponent.__docgenInfo = {
  props: {
    children: {
      description: 'support adding Section'
    }
  }
};

SectionListComponent.Item = SectionComponent;

export default SectionListComponent;
