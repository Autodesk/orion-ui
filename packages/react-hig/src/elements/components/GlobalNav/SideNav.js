/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import * as PropTypes from 'prop-types';
import createComponent from '../../../adapters/createComponent';
import HIGElement from '../../HIGElement';
import HIGChildValidator from '../../HIGChildValidator';

import SectionListComponent, { SectionList } from './SectionList';
import LinkListComponent, { LinkList } from './LinkList';
import SearchComponent, { Search } from './Search';

export class SideNav extends HIGElement {
  componentDidMount() {
    this.filter = this.filter.bind(this);

    if (this.sections) {
      this.sections.mount();
    }

    if (this.links) {
      this.links.mount();
    }

    if (this.search) {
      this.search.mount();
      this.search.hig.onInput(this.filter);
    }
  }

  createElement(ElementConstructor, props) {
    switch (ElementConstructor) {
      case SectionList:
        return new SectionList(this.hig); // special case hand over the hig instance
      case LinkList:
        return new LinkList(this.hig); // special case hand over the hig instance
      case Search:
        return new Search(this.hig.partials.Search, props);
      default:
        throw new Error(`Unknown type ${ElementConstructor.name}`);
    }
  }

  appendChild(instance) {
    if (instance instanceof SectionList) {
      if (this.sections) {
        throw new Error('only one SectionList is allowed');
      } else {
        this.sections = instance;

        if (this.mounted) {
          instance.componentDidMount();
        }
      }
    } else if (instance instanceof LinkList) {
      if (this.links) {
        throw new Error('only one LinkList is allowed');
      } else {
        this.links = instance;

        if (this.mounted) {
          instance.componentDidMount();
        }
      }
    } else if (instance instanceof Search) {
      if (this.search) {
        throw new Error('only one Search is allowed');
      } else {
        this.search = instance;

        if (this.mounted) {
          instance.componentDidMount();
          instance.hig.onInput(this.filter);
        }
      }
    } else {
      throw new Error('unknown type');
    }
  }

  insertBefore(instance, insertBeforeIndex) {
    this.appendChild(instance);
  }

  removeChild(instance) {
    if (instance instanceof SectionList) {
      this.sections = null;
    }

    if (instance instanceof LinkList) {
      this.links = null;
    }

    if (instance instanceof Search) {
      this.search = null;
    }

    instance.unmount();
  }

  filter(event) {
    const query = event.target.value;
    this.sections.forEach(section => {
      section.filter(query);
    });
  }
}

const SideNavComponent = createComponent(SideNav);

SideNavComponent.propTypes = {
  addSection: PropTypes.func,
  addLink: PropTypes.func,
  addSearch: PropTypes.func,
  setCopyright: PropTypes.func,
  children: HIGChildValidator([SectionListComponent, LinkListComponent, SearchComponent])
};

SideNavComponent.__docgenInfo = {
  props: {
    children: {
      description: 'support adding SectionList, and LinkList'
    }
  }
};

SideNavComponent.SectionList = SectionListComponent;
SideNavComponent.LinkList = LinkListComponent;
SideNavComponent.Search = SearchComponent

export default SideNavComponent;
