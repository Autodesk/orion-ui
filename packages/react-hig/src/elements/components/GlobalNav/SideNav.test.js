/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import GlobalNav from './GlobalNav';
import SideNav from './SideNav';
import Search from './Search';
import SectionList from './SectionList';
import Section from './Section';
import Group from './Group';
import Module from './Module';
import Submodule from './Submodule';

import LinkList from './LinkList';

describe('<SideNav>', () => {
  function createHigNav() {
    const higContainer = document.createElement('div');

    // use spread here to clone defaults since HIG.Button mutates this object
    const higNav = new HIG.GlobalNav();

    higNav.mount(higContainer);

    return { higNav, higContainer };
  }

  let Context = props => <GlobalNav>{props.children}</GlobalNav>;

  it('can render a SectionList as a child', () => {
    const { higNav, higContainer } = createHigNav();

    const sideNav = new higNav.partials.SideNav();
    higNav.addSideNav(sideNav);

    const reactContainer = document.createElement('div');
    const wrapper = mount(
      <GlobalNav>
        <SideNav>
          <SectionList />
        </SideNav>
      </GlobalNav>,
      {
        attachTo: reactContainer
      }
    );

    expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstChild.outerHTML).toEqual(
      higContainer.firstChild.outerHTML
    );
  });

  it('errors when there are multiple SectionList elements', () => {
    expect(() =>
      mount(
        <GlobalNav>
          <SideNav>
            <SectionList />
            <SectionList />
          </SideNav>
        </GlobalNav>
      )).toThrowError(/only one SectionList is allowed/);
  });

  it('can render a LinksList as a child', () => {
    const { higNav, higContainer } = createHigNav();

    const sideNav = new higNav.partials.SideNav();
    higNav.addSideNav(sideNav);

    const reactContainer = document.createElement('div');
    const wrapper = mount(
      <GlobalNav>
        <SideNav>
          <LinkList />
        </SideNav>
      </GlobalNav>,
      {
        attachTo: reactContainer
      }
    );

    expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();

    expect(reactContainer.firstChild.outerHTML).toEqual(
      higContainer.firstChild.outerHTML
    );
  });

  it('errors when there are multiple LinkList elements', () => {
    expect(() =>
      mount(
        <GlobalNav>
          <SideNav>
            <LinkList />
            <LinkList />
          </SideNav>
        </GlobalNav>
      )).toThrowError(/only one LinkList is allowed/);
  });

  it('can not render HTML elements as children', () => {
    global.console.error = jest.fn();

    mount(
      <GlobalNav>
        <SideNav>
          <div>Hello world!</div>
        </SideNav>
      </GlobalNav>
    );

    expect(console.error).toBeCalledWith(
      expect.stringMatching(/'div' is not a valid child of SideNav./)
    );
  });

  it('can not render text as children', () => {
    global.console.error = jest.fn();

    mount(
      <GlobalNav>
        <SideNav>
          Hello World!
        </SideNav>
      </GlobalNav>
    );

    expect(console.error).toBeCalledWith(
      expect.stringMatching(/'Hello World!' is not a valid child of SideNav./)
    );
  });

  describe('<SideNav> with search', () => {
    const Context = props => {
      return (
        <GlobalNav>
          <SideNav>
            <Search query={props.query} />
            <SectionList>
              <Section headerLabel="Project" headerName="Thunderstor">
                <Group>
                  <Module title={props.moduleTitle} icon="assets">
                    <Submodule title={props.submoduleTitle1} icon="assets" />
                    <Submodule title={props.submoduleTitle2} icon="assets" />
                  </Module>
                </Group>
              </Section>

            </SectionList>
          </SideNav>

        </GlobalNav>
      );
    };

    function setupReactContext(props) {
      const reactContainer = document.createElement('div');
      mount(<Context {...props} />, { attachTo: reactContainer });
      return { reactContainer };
    }

    it('renders children correctly', () => {
      const props = {
        placeholder: 'Search Here',
        headerLabel: 'Oakland Medical Center',
        headerName: 'Thunderstorm',
        moduleTitle: 'Document Workflow',
        submoduleTitle1: 'Document',
        submoduleTitle2: 'Workflow'
      };
      const { reactContainer } = setupReactContext(props);
      expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();
    });
  });
});
