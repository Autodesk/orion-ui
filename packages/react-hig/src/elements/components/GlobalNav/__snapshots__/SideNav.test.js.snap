// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNav> can render a LinksList as a child 1`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`<SideNav> can render a SectionList as a child 1`] = `
"<div class=\\"hig__global-nav\\">
    <div class=\\"hig__global-nav__sidenav\\">
    <!--SECTION-->
    <div class=\\"hig__global-nav__sidenav__links\\">
        <!--LINK-->
    </div>
    <div class=\\"hig__global-nav__sidenav__copywright\\">© 2017 Autodesk, Inc. </div>
</div><!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__container__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;
